### PR TITLE
Shenli/sql helper

### PR DESCRIPTION
**Pull request by shenli**
Make operations on system table easier.

### CODE REVIEW

File: stmt/stmts/account_manage.go [hunk 1]
```diff
@@ -73,39 +70,13 @@ func (s *CreateUserStmt) SetText(text string) {
...
-	p, err := r.Plan(ctx)
-	if err != nil {
-		return false, errors.Trace(err)
-	}
-	where := &rsets.WhereRset{
-		Src:  p,
-		Expr: composeUserTableFilter(name, host),
-	}
-	p, err = where.Plan(ctx)
+	sql := fmt.Sprintf("SELECT * FROM %s.%s WHERE User=\"%s\" AND HOST=\"%s\"", mysql.SystemDB, mysql.UserTable, name, host)
```
Comment (c4pt0r) [inline stmt/stmts/account_manage.go]: "AND HOST"
Are you sure?

File: session.go [hunk 2]
```diff
@@ -228,6 +230,33 @@ func (s *session) Retry() error {
...
+		log.Warnf("No statement to exec for %s", sql)
+	}
+	st := statements[0]
+	// Check statement for some restriction
+	// For example only support DML on system meta table.
+	log.Infof("Executing %s [%s]", st, sql)
+	ctx.SetValue(&sqlhelper.KeyType{}, true)
+	defer ctx.ClearValue(&sqlhelper.KeyType{})
+	rs, err := st.Exec(ctx)
+	if err != nil {
```
Comment (ngaut) [inline session.go]: Remove these 3 lines. and wrap to errors.Trace(err)

File: session.go [hunk 3]
```diff
@@ -228,6 +230,33 @@ func (s *session) Retry() error {
...
+	if ctx.Value(&sqlhelper.KeyType{}) != nil {
+		return nil, nil
+	}
+	statements, err := Compile(sql)
+	if err != nil {
+		log.Errorf("Syntax error: %s", sql)
+		return nil, errors.Trace(err)
+	}
+	if len(statements) == 0 {
+		log.Warnf("No statement to exec for %s", sql)
```
Comment (c4pt0r) [inline session.go]: missing return.

File: util/sqlhelper/sql_helper.go [hunk 4]
```diff
@@ -0,0 +1,20 @@
+package sqlhelper
```
Comment (c4pt0r) [inline util/sqlhelper/sql_helper.go]: add license header

File: stmt/stmts/account_manage.go (hunk 1 above)
Comment (ngaut) [inline stmt/stmts/account_manage.go]: using ` instead of "

File: stmt/stmts/account_manage.go [hunk 5]
```diff
@@ -136,22 +100,20 @@ func (s *CreateUserStmt) Exec(ctx context.Context) (rset.Recordset, error) {
...
 		if spec.AuthOpt.ByAuthString {
-			value = append(value, expression.Value{Val: util.EncodePassword(spec.AuthOpt.AuthString)})
+			pwd = util.EncodePassword(spec.AuthOpt.AuthString)
 		} else {
-			// TODO: Maybe we should hash the string here?
-			value = append(value, expression.Value{Val: util.EncodePassword(spec.AuthOpt.HashString)})
+			pwd = util.EncodePassword(spec.AuthOpt.HashString)
 		}
-		values = append(values, value)
+		user := fmt.Sprintf("(\"%s\", \"%s\", \"%s\")", host, userName, pwd)
```
Comment (ngaut) [inline stmt/stmts/account_manage.go]: using ` instead of "

File: stmt/stmts/account_manage.go [hunk 6]
```diff
@@ -188,20 +150,13 @@ func (s *SetPwdStmt) SetText(text string) {
...
-		ColName: "Password",
-		Expr:    expression.Value{Val: util.EncodePassword(s.Password)},
-	}
-	st := &UpdateStmt{
-		TableRefs: composeUserTableRset(),
-		List:      []expression.Assignment{asgn},
-		Where:     composeUserTableFilter(userName, host),
-	}
-	return st.Exec(ctx)
+	sql := fmt.Sprintf("UPDATE %s.%s SET password=\"%s\" WHERE User=\"%s\" AND Host=\"%s\";", mysql.SystemDB, mysql.UserTable, util.EncodePassword(s.Password), userName, host)
```
Comment (ngaut) [inline stmt/stmts/account_manage.go]: using ` instead of "

**Comment by shenli** [comment]
PTAL

File: session.go [hunk 7]
```diff
@@ -228,6 +230,36 @@ func (s *session) Retry() error {
 	return nil
 }
 
+// ExecRestrictedSQL implements SQLHelper interface.
+// This is used for execute some restricted sql statement.
```
Comment (qiuyesuifeng) [inline session.go]: is used for executing?
statement -> statements?

File: session.go [hunk 8]
```diff
@@ -228,6 +230,36 @@ func (s *session) Retry() error {
...
+func (s *session) ExecRestrictedSQL(ctx context.Context, sql string) (rset.Recordset, error) {
+	if ctx.Value(&sqlhelper.KeyType{}) != nil {
+		// We do not support run this function concurrently.
+		// TODO: Maybe we should remove this restriction latter.
+		return nil, errors.New("Should not call ExecRestrictedSQL concurrently.")
+	}
+	statements, err := Compile(sql)
+	if err != nil {
+		log.Errorf("Syntax error: %s", sql)
```
Comment (qiuyesuifeng) [inline session.go]: Need add error info in log?

File: session.go [hunk 9]
```diff
@@ -228,6 +230,36 @@ func (s *session) Retry() error {
...
+	// Check statement for some restriction
+	// For example only support DML on system meta table.
+	log.Infof("Executing %s [%s]", st, sql)
+	ctx.SetValue(&sqlhelper.KeyType{}, true)
+	defer ctx.ClearValue(&sqlhelper.KeyType{})
+	rs, err := st.Exec(ctx)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return rs, err
```
Comment (qiuyesuifeng) [inline session.go]: errors.Trace(err)?

File: stmt/stmts/account_manage.go (hunk 5 above)
Comment (qiuyesuifeng) [inline stmt/stmts/account_manage.go]: +1
user := fmt.Sprintf(`("%s", "%s", "%s")`, host, userName, pwd)

File: util/sqlhelper/sql_helper.go [hunk 10]
```diff
@@ -0,0 +1,37 @@
+// Copyright 2013 The ql Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSES/QL-LICENSE file.
```
Comment (qiuyesuifeng) [inline util/sqlhelper/sql_helper.go]: ql license can be removed?

File: session.go (hunk 9 above)
Comment (shenli) [inline session.go]: PTAL

File: util/sqlhelper/sql_helper.go (hunk 10 above)
Comment (shenli) [inline util/sqlhelper/sql_helper.go]: PTAL

**Comment by c4pt0r** [comment]
LGTM

File: util/sqlhelper/sql_helper.go [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+// KeyType is a dummy type to avoid naming collision in session.
+type KeyType struct{}
+
+// String implements Stringer interface.
+func (k *KeyType) String() string {
+	return "sql_helper"
+}
+
+// SQLHelper is an interface provides executing restricted sql statement.
+type SQLHelper interface {
```
Comment (ngaut) [inline util/sqlhelper/sql_helper.go]: Need to figure out a better name than SQLHelper

**Pull request merged**